### ISSUE TITLE
create "user-installable-sdk" version of KendoUI bridge

### ISSUE THREAD

**Issue by adriatic** [aurelia-ui-toolkits/aurelia-kendoui-bridge#556]
I took the time to define this task as precisely as I can; that way we could quickly zoom into the sections of interest, without loosing the overall context.

**Related issues**
- https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/issues/528
- https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/issues/552
- https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/issues/550

**Comment by adriatic** [comment]
### Task definition

Instead of having the bridge tasked with accessing KendoUI SDK widgets, we want to leave this task to the application developer. This change allows the application developer to **define the set of KendoUI widgets** which the bridge subsequently converts to Aurelia components. As a result, the choice between **core** set of widgets or **pro** set of widgets (the only two alternatives available so far, defined in via **[this line](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/sample/src/main.js#L9)** becomes a lot more fine-grained including all subsets of the set of KendoUI widgets.

**Comment by adriatic** [comment]
### Current implementation

___This description focuses only on the process of wrapping up a single KendoUI widget into the functionally equivalent Aurelia component.___

I will use the **[tabstrip](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/src/tabstrip/tabstrip.js)** as the model to describe this process that takes place when the application using the bridge -like this

```
      <div id="tabstrip" ak-tabstrip="k-animation.bind: { open: { effects: 'fadeIn' } }">
        <ul>
          <li class="k-state-active">
            Paris
          </li>
          <li>
          ...
```

The ___appearance___ of the keyword `ak-tabstrip` results with the invocation of bridge's wrapper for the KendoUI tabstrip widget defined in the `kendo.tabstrip.min` JavaScript file. Here is the wrapper's implementation

```
import {inject} from 'aurelia-dependency-injection';
import {customAttribute} from 'aurelia-templating';
import {WidgetBase} from '../common/widget-base';
import {generateBindables} from '../common/decorators';
import {constants} from '../common/constants';
import 'kendo.tabstrip.min';                                               <=========

@customAttribute(`${constants.attributePrefix}tabstrip`)
@generateBindables('kendoTabStrip')
@inject(Element, WidgetBase)
export class TabStrip {

  constructor(element, widgetBase) {
    this.element = element;
    this.widgetBase = widgetBase
                        .control('kendoTabStrip')
                        .linkViewModel(this);
  }

// the rest of the class is skipped 
}

```

The salient points are:

1. KendoUI widget is accessed in the wrapper via `import 'kendo.tabstrip.min';`
2. It is tagged as Aurelia custom attribute: @customAttribute(`${constants.attributePrefix}tabstrip`) 
3. All (previously determined) widget attributes are added as "class bindables". This step needs to be expanded for clarity:
 - all bindables are manually created and then defined in the [bindables.js](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/src/common/bindables.js) class. The set of tabstrip attributes is defined as 

   ```
   "kendoTabStrip":
   [
       "animation","collapsible","contentUrls","dataContentField",
       "dataContentUrlField","dataImageUrlField","dataSource",
       "dataSpriteCssClass","dataTextField","dataUrlField","name",
       "navigatable","scrollable","tabPosition","value"
   ]
   ```
4. The actual conversion from the "plain javascript class" into an Aurelia component, takes place on the wrapper's `constructor`
 
 ```
  constructor(element, widgetBase) {
    this.element = element;
    this.widgetBase = widgetBase
                        .control('kendoTabStrip')
                        .linkViewModel(this);
 ```

 where the most interesting action takes place in the `control`method of the **[widget-base.js](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/src/common/widget-base.js)** class.

  ```
 control(controlName) {
    if (!controlName || !kendo.jQuery.fn[controlName]) {
      throw new Error(`The name of control ${controlName} is invalid or not set`);
    }

    this.controlName = controlName;

    let ctor = kendo.jQuery.fn[this.controlName];
    this.kendoOptions = ctor.widget.prototype.options;
    this.kendoEvents = ctor.widget.prototype.events;

    return this;
  }
  ```

**Comment by adriatic** [comment]
### Summary of the current implementation

1. Wrapper has the "hard-coded" access to the KendoUI widget: `import 'kendo.tabstrip.min'
2. Wrapper invokes the `control` method of the `widget-base.js` class passing the hardcoded string 'kendoTabStrip' as the argument, "knowing" that this name ___should be found___ in the `kendo.tabstrip.min.js` class with implements the widget.

The rest of the processing is not relevant for this discussion.

**Comment by adriatic** [comment]
### How to move forward from here

The ideal solution would to simply provide the wrapper with the access to the 'kendo.tabstrip.min' class, ensuring that the rest of the processing takes place in the bridge as it is now. Consider the simplest possible situation where the app developer wants only tabstrip aurelia component.

**Comment by adriatic** [comment]
Just pushed the user-installable-sdk branch (55463b1015599cf542dcfb0407890bbe6720192b). It does not work of course

**Comment by JeroenVinke** [comment]
> the application tells the bridge that it want to use the set of core or pro widgets. This information is used where ???

Nowhere really, we just make it easy for the user to load all of our kendo wrappers (and because of the `import kendo.x.min.js` statements in each wrapper it also loads all kendo controls). We're not checking anywhere if you're loading Kendo CORE or Kendo PRO

> I am still not sure whether each widget is "processed" at the time it is needed by the application or all processing takes place at init time; I believe it is the first variant

This depends, if you use `kendo => kendo.pro()` in main.js then all wrappers and subsequent Kendo controls will be processed at application startup. If you don't do this, and add `<require from="aurelia-kendoui-bridge/tabstrip/tabstrip"></require>` in a view then the tabstrip wrapper and control will be processed when the view is used for the first time. The latter is a much better approach IMO. Imagine loading all kendo controls at application startup and the user never reaching a screen where a kendo control is used. All that processing at application startup would have been for nothing. When you take into consideration that `kendo => kendo.pro()` can sometimes delay startup time of the app by 5 seconds then using `kendo => kendo.pro()` as "default" in our installation instructions can be labelled as bad advice

https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/commit/55463b1015599cf542dcfb0407890bbe6720192b is good. If you agree with my view on how bad `kendo => kendo.core()` and `kendo => kendo.pro()` is, then we should remove this "feature" from our bridge and steer users towards using `<require>` in their views where they need a kendo control.

Also, on a high level there seems to be two ways to load Kendo controls. What you did ([here](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/commit/55463b1015599cf542dcfb0407890bbe6720192b#diff-3e8cb08b12758b91cb32438656c9413bR22)), by loading Kendo controls via index.html (great if you would like to use kendo.all.min.js or kendo.custom.min.js), but we should confirm if users can also load kendo controls via a module loader by adding `import 'kendo.tabstrip.min';` in the view-model where a Kendo control is used. Ask yourself this: why download a 2MB+ payload of Kendo PRO controls when you don't even know if the user is ever going to need it (they might not reach a screen where there is a Kendo control). The trick with the module loader approach is to have a Kendo control loaded **before** our wrapper [initializes](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/55463b1015599cf542dcfb0407890bbe6720192b/src/tabstrip/tabstrip.js#L31) the control.

This is a complete guess, but you could try this out:

```
<require from="kendo.tabstrip.min"></require>
<require from="aurelia-kendoui-bridge/tabstrip/tabstrip"></require>
// use tabstrip here
```

or


```
// in the view
<require from="aurelia-kendoui-bridge/tabstrip/tabstrip"></require>
// use tabstrip here

// in the view-model
import 'kendo.tabstrip.min';
```

This would create the following installation instruction flow:

- via index.html
- via module loader
  - jspm
    - core
    - pro
    - vendors
  - webpack
    - core
    - pro
    - vendors
  - requirejs
    - core
    - pro
    - vendors

**Comment by adriatic** [comment]
> If you agree with my view on how bad kendo => kendo.core() and kendo => kendo.pro() is, then we should remove this "feature" from our bridge and steer users towards using <require> in their views where they need a kendo control.

I agree with this wholeheartedly - it does nothing good for us and adds to the initial load time.

> we should confirm if users can also load kendo controls via a module loader by adding import 'kendo.tabstrip.min'; in the view-model where a Kendo control is used. 

I am on board with that completely with the caveat:

1. Assume that the user does have the access to `kendo-tabstrip-min` (this is true when using vendors, or kendo-core via bower). How will that file be "passed" to the respective wrapper? One could say it does not need to as the wrapper already has hardcoded path to that file

2. The problems start with user having the `kendo.custom.min`, while the bridge expects a separate .min file for each widget (or, non minified file). Can we take the `kendo.custom.min` file apart and stuff the result in an agreed upon location so that the bridge finds it as if we used the `vendors` approach>

**Comment by JeroenVinke** [comment]
> How will that file be "passed" to the respective wrapper?

Let's look at how kendo controls are initialized without the bridge: `$("<div></div>").kendoButton()`. The `kendoButton()`  function exists on the jquery global object `$`. The `$` variable is available everywhere without having to import anything first. And there is only one `$` (assuming there are no forks). When you load `kendo.button.min.js` it registers the `kendoButton()` function on the jquery object (`$`). So after the kendo.button.min.js has been loaded (it doesn't matter when or by who) you can use it anywhere in the system via `$("<div></div>")`, so we can use it in our Button wrapper without having to import it. We just assume that a `kendoButton()` function exists on the jquery object.

We now have `import` statements in our wrappers. This tells a module loader to import a file, for example `kendo.button.min.js`. This way we **know** that the `kendoButton()` function is on the jQuery object, and we don't have to assume. But if we remove this import, and the wrapper assumes that the kendoButton() function is available, then you can load `kendo.button.min.js` (or `kendo.custom.min.js` if it contains the button control) yourself and our bridge would work just fine.

**Comment by charlespockert** [comment]
We can't really do conditional imports can we? If we could that would mean we didn't have to worry about what loader was present... I still like the idea of the loader pulling in individual files (I.e. just what you need)

In fact for smaller apps using only a couple of controls it's not that bad loading everything up front

**Comment by adriatic** [comment]
> In fact for smaller apps using only a couple of controls it's not that bad loading everything up front

The word "everything" means everything that the app needs - not everything that KendoUI has to offer, right?

**Comment by JeroenVinke** [comment]
what about

```
// button-wrapper.min.js
import 'kendo.button.min';
import './button.js';
```

where `./button.js`  is our wrapper (that doesn't import kendo.button.min.js) and could be loaded instead of `button-wrapper.min.js`

**Comment by adriatic** [comment]
> where ./button.js is our wrapper (that doesn't import kendo.button.min.js) 

This looks cool - assuming that this jQuery plugin concept is hiddent in it.

**Comment by adriatic** [comment]
Please sketch me the code in the `app` (the `bridge` configuration in `main.js` and the access to a single kendo widget `kendoButton.min.js`)   and the wrapper for that widget in  the `bridge `). In the meantime I will read-up more on jQuery plugins to be able to understand how to **put** the reference to `kendoButton.min.js` in the "query object" in the context of `app` and how to **get** it in the context of the  **[wrapper](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/src/button/button.js)** in the bridge which now does not have the `import 'kendo.button.min';` statement.

**Comment by JeroenVinke** [comment]
Please make sure that you understand the following as it is important for the rest of my story:
```
<html>
  <head>
     // $ is undefined
     <script src="http://code.jquery.com/jquery-1.9.1.min.js"></script>
     // $ is defined, $().kendoButton function is undefined
    <script src="http://kendo.cdn.telerik.com/2016.2.607/js/kendo.button.min.js"></script>
    // $.kendoButton is defined (happened because kendo.button.min.js was loaded, and kendo.button.min.js registers the kendoButton function on the global $ object)
    // Since $ is a global, we can use $.kendoButton throughout the system without having to load kendo.button.min.js again. 
   // We do not need to pass $.kendoButton to our wrappers, the wrapper can just call $.kendoButton() since $ is a global
  </head>
  </head>
  <body>
  </body>
</html>
```

What you did [here](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/commit/55463b1015599cf542dcfb0407890bbe6720192b) is correct and enough, let me explain.

Let's assume that we are **not** loading Kendo from index.html, and that our wrapper is called 'button-wrapper.js' and it has an `import` statement:
```
// button-wrapper.js
import 'kendo.button.min.js';

export class ButtonWrapper () {
  recreate() {
    $(this.element).kendoButton();
  }
}
```

When you tell any module loader (systemjs/requirejs) to load `kendo-wrapper.js` it will see that `import` statement and it will load `kendo.button.min.js` **before** loading `kendo-wrapper.js`. When you're then calling `recreate()`, you know that `kendo.button.min.js` has already been loaded at that point. And if you remember from the HTML code in the beginning, after `kendo.button.min.js` has been loaded once the `$.kendoButton()` function is ready to be used.

Now, when you remove `import 'kendo.button.min.js'` and instead load `kendo.button.min.js` from your index.html, `$.kendoButton()` is still available when instantiating `ButtonWrapper`, since `$` is a global and is available everywhere. So even without `import` statements, we can still use `$.kendoButton()`.

Conclusion: We can remove all `import` statements from our wrapper, and from that moment we assume that `kendo.button.min.js` has been loaded (in other words, we assume that the `$.kendoButton()` function exists) before our wrappers are used. 

It does not matter how the user is going to load `kendo.button.min.js`, they can load it through index.html: 
- `<script src="http://kendo.cdn.telerik.com/2016.2.607/js/kendo.button.min.js"></script>`.
- `<script src="http://kendo.cdn.telerik.com/2016.2.607/js/kendo.all.min.js"></script>`.
- `kendo.custom.min.js"></script>`.

Or they can load it through a module loader by including this import statement in their app: `import 'kendo.button.min.js'`. Note that this import statement has moved from our wrapper to the users application, allowing them to **not** use a module loader by not using this `import statement`.

**Comment by adriatic** [comment]
> Or they can load it through a module loader by including this import statement in their app: import 'kendo.button.min.js'. Note that this import statement has moved from our wrapper to the users application, allowing them to not use a module loader by simply not using this import statement.

A very succinct definition of the solution for dramatically improving the KendoUI bridge (any bridge for that matter). At this point, I am surprised that we did not do that from the start - but as they say hindsight vision is always 20/20.

My gain in this discussion is huge as I never realized how important rule of jQuery is even today. Of course, it's not jQuery alone, as the loader (entity that processes all import statements or html files has to check whether jQuery is already present, and add **all** references to the `q` object. While I knew that our solution needs to do something like that, I had no idea that jQuery is already doing it. 

Thanks again for [such carefully written explanation](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/issues/556#issuecomment-231667004)

**Comment by adriatic** [comment]
Returning back after some fights with webpack and monterey book writing, with the desire to finish this off. So, I will pick up this conversation at your paragraph

> Conclusion: We can remove all `import` statements from our wrapper, and from that moment we assume that `kendo.button.min.js` has been loaded (in other words, we assume that the `$.kendoButton()` function exists) before our wrappers are used. 

>It does not matter how the user is going to load `kendo.button.min.js`, they can load it through index.html: 
- `<script src="http://kendo.cdn.telerik.com/2016.2.607/js/kendo.button.min.js"></script>`.
- `<script src="http://kendo.cdn.telerik.com/2016.2.607/js/kendo.all.min.js"></script>`.
- `<script src="kendo.custom.min.js"></script>`.


in this [section](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/issues/556#issuecomment-231667004)

I just checked in the needed changes (in `user-installable-sdk` branch that takes the simplest approach by loading the **[needed](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/user-installable-sdk/sample/index.html#L21-L23)** `kendo .min.js` files from **[index.html](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/user-installable-sdk/sample/index.html)** file

We already agreed that all  wrapper files in the **[src](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/tree/user-installable-sdk/src)** folder are set correctly, so we should be good to go. 

The `catalog` app uses the **[about](http://aurelia-ui-toolkits.github.io/demo-kendo/#/about/about)** page as the landing page, so the command `gulp watch` will first "hit" the tabstrip widget

![image](https://cloud.githubusercontent.com/assets/2712405/16897593/20e071c6-4b84-11e6-84ef-653bf0038317.png)

The console log is
```
DEBUG [aurelia] Loading plugin aurelia-templating-binding.
aurelia-logging-console.js:24 DEBUG [aurelia] Configured plugin aurelia-templating-binding.
aurelia-logging-console.js:24 DEBUG [aurelia] Loading plugin aurelia-templating-resources.
aurelia-logging-console.js:24 DEBUG [aurelia] Configured plugin aurelia-templating-resources.
aurelia-logging-console.js:24 DEBUG [aurelia] Loading plugin aurelia-history-browser.
aurelia-logging-console.js:24 DEBUG [aurelia] Configured plugin aurelia-history-browser.
aurelia-logging-console.js:24 DEBUG [aurelia] Loading plugin aurelia-templating-router.
aurelia-logging-console.js:24 DEBUG [aurelia] Configured plugin aurelia-templating-router.
aurelia-logging-console.js:24 DEBUG [aurelia] Loading plugin aurelia-event-aggregator.
aurelia-logging-console.js:24 DEBUG [aurelia] Configured plugin aurelia-event-aggregator.
aurelia-logging-console.js:24 DEBUG [aurelia] Loading plugin aurelia-kendoui-bridge.
aurelia-logging-console.js:24 DEBUG [aurelia] Configured plugin aurelia-kendoui-bridge.
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/shared/collapse-panel.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/shared/logger.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/autocomplete/autocomplete.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/calendar/calendar.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/colorpalette/colorpalette.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/combobox/combobox.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/dropdownlist/dropdownlist.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/flatcolorpicker/flatcolorpicker.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/listview/listview.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/multiselect/multiselect.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/notification/notification.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/panelbar/panelbar.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/rangeslider/rangeslider.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/scrollview/scrollview.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/toolbar/toolbar.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/toolbar/toolbar-item.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/toolbar/toolbar-item-button.html []
aurelia-logging-console.js:34 INFO [aurelia] Aurelia Started
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/app.html ["http://localhost:3000/src/nav-bar", "bootstrap/css/bootstrap.css"]
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/nav-bar.html ["theme-selector/theme-selector"]
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/theme-selector/theme-selector.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/about/about.html ["http://localhost:3000/src/about/about.css"]
aurelia-logging-console.js:54 ERROR [app-router] Error: Error invoking TabStrip. Check the inner error for details.(…)
```
<br>

where the inner error is 
```
------------------------------------------------
Inner Error:
Message: The name of control kendoTabStrip is invalid or not set
Inner Error Stack:
Error: The name of control kendoTabStrip is invalid or not set
    at WidgetBase.control (http://localhost:3000/src/aurelia-kendoui-bridge/common/widget-base.js:52:19)
    at new TabStrip (http://localhost:3000/src/aurelia-kendoui-bridge/tabstrip/tabstrip.js:32:40)
    at Object.invoke (http://localhost:3000/jspm_packages/npm/aurelia-dependency-injection@1.0.0-rc.1.0.0/aurelia-dependency-injection.js:356:14)
    at InvocationHandler.invoke (http://localhost:3000/jspm_packages/npm/aurelia-dependency-injection@1.0.0-rc.1.0.0/aurelia-dependency-injection.js:321:168)
    at Container.invoke (http://localhost:3000/jspm_packages/npm/aurelia-dependency-injection@1.0.0-rc.1.0.0/aurelia-dependency-injection.js:585:25)
    at ProviderResolver.get (http://localhost:3000/jspm_packages/npm/aurelia-templating@1.0.0-rc.1.0.0/aurelia-templating.js:2013:74)
    at Container.get (http://localhost:3000/jspm_packages/npm/aurelia-dependency-injection@1.0.0-rc.1.0.0/aurelia-dependency-injection.js:522:23)
    at Container.elementContainerGet [as get] (http://localhost:3000/jspm_packages/npm/aurelia-templating@1.0.0-rc.1.0.0/aurelia-templating.js:2068:17)
    at HtmlBehaviorResource.create (http://localhost:3000/jspm_packages/npm/aurelia-templating@1.0.0-rc.1.0.0/aurelia-templating.js:3961:58)
    at applyInstructions (http://localhost:3000/jspm_packages/npm/aurelia-templating@1.0.0-rc.1.0.0/aurelia-templating.js:2162:33)
    at ViewFactory.create (http://localhost:3000/jspm_packages/npm/aurelia-templating@1.0.0-rc.1.0.0/aurelia-templating.js:2378:9)
    at HtmlBehaviorResource.create (http://localhost:3000/jspm_packages/npm/aurelia-templating@1.0.0-rc.1.0.0/aurelia-templating.js:3973:41)
    at http://localhost:3000/jspm_packages/npm/aurelia-templating-router@1.0.0-rc.1.0.0/router-view.js:163:51
    at r (https://cdnjs.cloudflare.com/ajax/libs/bluebird/3.4.0/bluebird.min.js:31:7722)
    at r._settlePromiseFromHandler (https://cdnjs.cloudflare.com/ajax/libs/bluebird/3.4.0/bluebird.min.js:30:13014)
    at r._settlePromise (https://cdnjs.cloudflare.com/ajax/libs/bluebird/3.4.0/bluebird.min.js:30:13817)
End Inner Error Stack
```

So, it claims **`The name of control kendoTabStrip is invalid or not set`** at **`widget-base.js:52:19`**

however the debugger claims differently (all seems fine there). This likely means that the debugger is somehow confused - or ___I am confused even more.___


![image](https://cloud.githubusercontent.com/assets/2712405/16897715/8d93ab3c-4b87-11e6-9d2c-cde6d5f9b7ab.png)

**Comment by JeroenVinke** [comment]
Ah, classic mistake. kendo.core.min.js is not the Kendo CORE suite, it is the base class from which all controls inherit. Going off of [this page](http://docs.telerik.com/kendo-ui/intro/installation/cdn-service) you need `kendo.ui.core.min.js`. However, now that the bridge is no longer importing anything, you should be able to include `kendo.tabstrip.min.js` as well

**Comment by adriatic** [comment]
I might be down to the last "unexplainable" problem: at the point I felt being done with this task, running `gulp watch` results with:
```
...
DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/panelbar/panelbar.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/rangeslider/rangeslider.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/scrollview/scrollview.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/toolbar/toolbar.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/toolbar/toolbar-item.html []
aurelia-logging-console.js:24 DEBUG [templating] importing resources for http://localhost:3000/src/aurelia-kendoui-bridge/toolbar/toolbar-item-button.html []
aurelia-logging-console.js:34 INFO [aurelia] Aurelia Started
css.js:56 GET http://localhost:3000/src/root/vendors/styles/kendo.common.min.css (anonymous function) @ css.js:56e._execute @ bluebird.min.js:29r._resolveFromExecutor @ bluebird.min.js:30r @ bluebird.min.js:30loadCSS @ css.js:28exports.fetch @ css.js:65(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597r @ bluebird.min.js:31r._settlePromiseFromHandler @ bluebird.min.js:30r._settlePromise @ bluebird.min.js:30r._settlePromise0 @ bluebird.min.js:30r._settlePromises @ bluebird.min.js:30r._drainQueue @ bluebird.min.js:29r._drainQueues @ bluebird.min.js:29drainQueues @ bluebird.min.js:29
css.js:56 GET http://localhost:3000/src/root/vendors/styles/kendo.default.min.css (anonymous function) @ css.js:56e._execute @ bluebird.min.js:29r._resolveFromExecutor @ bluebird.min.js:30r @ bluebird.min.js:30loadCSS @ css.js:28exports.fetch @ css.js:65(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597r @ bluebird.min.js:31r._settlePromiseFromHandler @ bluebird.min.js:30r._settlePromise @ bluebird.min.js:30r._settlePromise0 @ bluebird.min.js:30r._settlePromises @ bluebird.min.js:30r._drainQueue @ bluebird.min.js:29r._drainQueues @ bluebird.min.js:29drainQueues @ bluebird.min.js:29
css.js:56 GET http://localhost:3000/src/root/vendors/styles/kendo.default.mobile.min.css (anonymous function) @ css.js:56e._execute @ bluebird.min.js:29r._resolveFromExecutor @ bluebird.min.js:30r @ bluebird.min.js:30loadCSS @ css.js:28exports.fetch @ css.js:65(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597r @ bluebird.min.js:31r._settlePromiseFromHandler @ bluebird.min.js:30r._settlePromise @ bluebird.min.js:30r._settlePromise0 @ bluebird.min.js:30r._settlePromises @ bluebird.min.js:30r._drainQueue @ bluebird.min.js:29r._drainQueues @ bluebird.min.js:29drainQueues @ bluebird.min.js:29
bluebird.min.js:29 Unhandled rejection Error: Error loading CSS file.
    at HTMLLinkElement.link.onerror (http://localhost:3000/jspm_packages/github/systemjs/plugin-css@0.1.20/css.js:54:34)
```

One of the changes I did was to remove the `vendors folder` completely and added the `kendo-sdk` folder in `aurelia-kendoui-bridge/sample` - the root of the catalog application (as this is where this should live)

Note the failure to load `http://localhost:3000/src/root/vendors/styles/kendo.default.min.css ` - meaning that this is somehow hard coded somewhere (bluebird ??), since I am certain that the bridge code as it is currently pushed to the `user-installable-sdk` branch has no references to `vendors/styles` anything.

It seems that you will get the chance to see how this new version works first :-) - and I am quite curious what stops me here!!

**Comment by JeroenVinke** [comment]
This is where our catalog differs from normal applications using the bridge. We have the theme-manager loading styles via [SystemJS](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/sample/src/shared/theme-manager.js#L70) and in normal application you would have one theme and load it with `<require from="kendo-ui/styles/kendo.bootstrap.min.css"></require>` or via `<link rel="stylesheet" href="https://kendo.cdn..../kendo.bootstrap.min.css/>`. 

Note that the theme-manager is using `kendo-ui` which is a path mapping in [config.js](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/sample/config.js#L17). You can remove this path mapping and change [this line](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/sample/src/shared/theme-manager.js#L70) into `return System.normalizeSync(`/kendo-sdk/styles/kendo.${theme}.min.css!`);`. Also, you need to remove [this stylesheet](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/user-installable-sdk/sample/index.html#L9) for it to work, because it should be loaded by the theme manager.

Also, there are some imports left that need to be removed:
![image](https://cloud.githubusercontent.com/assets/2189477/16908179/fb8608a4-4cc9-11e6-8793-3ea907f82520.png)
![image](https://cloud.githubusercontent.com/assets/2189477/16908204/16739a50-4cca-11e6-9658-fc6669994725.png)

**Comment by adriatic** [comment]
> Also, there are some imports left that need to be removed:

I have no explanations for missing these imports (explanations to myself I mean). I should have remembered the issues with the theme-manager as well. Inexcusable - no food today!

**Comment by adriatic** [comment]
This was a tough day - started late and spent all the time chasing jQuery references. After addressing the few "remove imports" above, I found a few million of places with `$(` or `$.` - you can see them all un this last push I am about to do now.

Despite the pain, I am happy that I asked to do this task - as the code is now a lot less sensitive to having multiple instances of jQuery loaded. Once I catch all remaining problems (all of them related to the sample rendering in `showcase.html`) this will be a better bridge :-)

So, tomorrow another day of debugging - if you feel some compassion, see if you can tell me why is grid not showing anything, as that might be a cue for other similar cases. I am too tired now to do it today

**Comment by adriatic** [comment]
Added few more $ --> kendo.jQuery replacements (at important places), but many of the samples are still misbehaving (Treeview is a good example, besides grid)

**Comment by adriatic** [comment]
Just checked in all the changes I was planning to do in this round:

![image](https://cloud.githubusercontent.com/assets/2712405/16954258/36d1cb3a-4d9e-11e6-9a6c-e8c5ba7b5485.png)

Note that the version number is one higher than the last released version number.

**Comment by JeroenVinke** [comment]
closing this, as it is in the beta branch and released as 1.0.0-beta1.0.0 (prerelease)
